File: src/waiver.ts
import { readFileSync } from "node:fs";
import { normalizeSection } from "./config.js";
import { globMatches, normalizePathForMatch } from "./glob.js";
import type { Finding } from "./model.js";

interface Waiver {
  id?: string;
  rules?: string[];
  allow?: string[];
  scope?: {
    path?: string;
    doc_type?: string;
    section?: string;
  };
  expires?: string;
}

export function applyWaivers(findings: Finding[], waiverFile?: string): Finding[] {
  if (!waiverFile) {
    return findings;
  }
  const waivers = loadWaivers(waiverFile);
  const today = new Date().toISOString().slice(0, 10);
  return findings.map((finding) => {
    const waiverId = matchingWaiver(finding, waivers, today);
    return waiverId ? { ...finding, waivedBy: waiverId } : finding;
  });
}

function loadWaivers(path: string): Waiver[] {
  const data = JSON.parse(readFileSync(path, "utf8")) as unknown;
  if (Array.isArray(data)) {
    return data as Waiver[];
  }
  if (data && typeof data === "object" && Array.isArray((data as { waivers?: unknown }).waivers)) {
    return (data as { waivers: Waiver[] }).waivers;
  }
  throw new Error("waiver file must contain a list or a top-level 'waivers' list");
}

function matchingWaiver(finding: Finding, waivers: Waiver[], today: string): string | undefined {
  for (const waiver of waivers) {
    const waiverId = waiver.id ?? "";
    if (!waiverId) {
      continue;
    }
    if (waiver.expires && waiver.expires < today) {
      continue;
    }
    const rules = waiver.rules ?? waiver.allow ?? [];
    if (!rules.includes(finding.rule) && !rules.includes("*")) {
      continue;
    }
    if (!scopeMatches(finding, waiver.scope ?? {})) {
      continue;
    }
    return waiverId;
  }
  return undefined;
}

function scopeMatches(finding: Finding, scope: NonNullable<Waiver["scope"]>): boolean {
  if (scope.path) {
    const path = normalizePathForMatch(finding.path);
    const pattern = normalizePathForMatch(scope.path);
    if (!globMatches(path, pattern)) {
      return false;
    }
  }

  if (scope.doc_type && scope.doc_type !== finding.docType) {
    return false;
  }

  if (scope.section) {
    const expected = normalizeSection(scope.section);
    if (!finding.sections.some((section) => normalizeSection(section) === expected)) {
      return false;
    }
  }

  return true;
}

File: markdown_gate/waiver.py
from __future__ import annotations

from dataclasses import replace
from datetime import date
import fnmatch
import json
from pathlib import Path
from typing import Any

from .config import normalize_section
from .model import Finding


def apply_waivers(findings: list[Finding], waiver_file: str | None) -> list[Finding]:
    if not waiver_file:
        return findings

    waivers = _load_waivers(Path(waiver_file))
    today = date.today()
    updated: list[Finding] = []
    for finding in findings:
        waiver_id = _matching_waiver(finding, waivers, today)
        updated.append(replace(finding, waived_by=waiver_id) if waiver_id else finding)
    return updated


def _load_waivers(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        raise FileNotFoundError(f"waiver file not found: {path}")
    data = json.loads(path.read_text(encoding="utf-8"))
    if isinstance(data, list):
        return data
    waivers = data.get("waivers", [])
    if not isinstance(waivers, list):
        raise ValueError("waiver file must contain a list or a top-level 'waivers' list")
    return waivers


def _matching_waiver(
    finding: Finding,
    waivers: list[dict[str, Any]],
    today: date,
) -> str | None:
    for waiver in waivers:
        waiver_id = str(waiver.get("id") or "")
        if not waiver_id:
            continue

        expires = waiver.get("expires")
        if isinstance(expires, str) and expires:
            try:
                if date.fromisoformat(expires) < today:
                    continue
            except ValueError:
                continue

        rules = waiver.get("rules") or waiver.get("allow") or []
        if finding.rule not in rules and "*" not in rules:
            continue

        scope = waiver.get("scope") or {}
        if not _scope_matches(finding, scope):
            continue

        return waiver_id
    return None


def _scope_matches(finding: Finding, scope: dict[str, Any]) -> bool:
    path_pattern = scope.get("path")
    if path_pattern and not fnmatch.fnmatch(finding.path, str(path_pattern)):
        return False

    doc_type = scope.get("doc_type")
    if doc_type and str(doc_type) != finding.doc_type:
        return False

    section = scope.get("section")
    if section:
        if not finding.section:
            return False
        if normalize_section(str(section)) != normalize_section(finding.section):
            return False

    return True
